### ISSUE TITLE
Remove/separate newRef?

### ISSUE THREAD

**Issue by AriFordsham** [mainland/ref-tf#5]
Proposing removing the `newRef` method from the class, or separating it out into a separate class.

# Con:

- Standard unparameterised `newRef` is useful for (polymorphic) actions which use multiple variables.

# Pro:

- Some use cases for persistent storage may need extra parameters to create/initialise a variable, e.g. file-backed storage, databases, cloud storage. It would be useful to read/write/modify through this interface, but an unparameterised creation action is not appropriate for them. Rather, initialised instances of `Ref` can be passed into actions.

A 'best-of-both-worlds' approach would be to separate out `newRef` into a separate class, at the cost of possibly excessive complexity.

**Comment by mainland** [comment]
Do you have a concrete example in mind? Or, even better, a concrete use case? :)

**Comment by AriFordsham** [comment]
Yes. Imagine a FileMonad with FileRefs, each referring to a single file containing a single serialized ADT.

In this case, the following would make sense:

```haskell
sayHello :: (MonadRef m) => (Ref m) String -> m ()
sayHello ref = do
    writeRef ref "Hello, World!"

main :: IO ()
main = do
    file <- openFileRef "~/hello.hson"
    runFile $ sayHello file
```

However, the following wouldn't:

```haskell
sayHello :: (MonadRef m) => m ()
sayHello = do
    ref <- newRef "Hello, World!" -- Which file? Create or open? Overwrite or append?
    return ()

main :: IO ()
main = do
    runFile $ sayHello
```

I am doing extensive work with similar ideas. Your code is not hard to re-implement for my needs. My main motivations are a) to improve your project and b) to have this distinction on the Internet somewhere.

**Comment by mainland** [comment]
What if you wanted to have multiple serialized ADTs in the same file?

Here's an alternative: make a `FileMonad` whose actions can only be run in the context of a particular file. For example:

```haskell
runFileMonad :: FileMonad m => FilePath -> m a -> IO a
``` 

or something like that. You can now write a `MonadRef` instance for `FileMonad` since `newRef` will only be run in the context of a particular file.

**Comment by AriFordsham** [comment]
The idea is a good one.

My point was that supporting creation of an unparameterised reference is not a neccesary property of  Ref type/Monad, and therefore the code (At least in concept) would be more general if it was left out.

Again, it's not causing me any difficulty, I just wanted to have that on the record, and to bring it to your attention @mainland in case you felt it was worth acting upon.

I am emphasise conceptual purity at the expense of (immediate) ease-of-use, so it's reasonable if it is not a good trade-off for your library.
